File: tools/tirtc-api-client/nodejs/tirtc_signing.js
/**
 * TGV1-HMAC-SHA256 request signing for tange.ai server APIs.
 *
 * Usage:
 *   const { signRequest } = require('./tirtc_signing');
 *   const headers = signRequest(accessKey, accessSecret, appId, 'POST', '/v2/user/login/user-id', '{"user_id":"test"}');
 */

const crypto = require('crypto');

const ALG_TGV1 = 'TGV1-HMAC-SHA256';

/**
 * @param {string} accessKey
 * @param {string} accessSecret
 * @param {string} appId
 * @param {string} method
 * @param {string} uriPath
 * @param {string} rawQuery
 * @param {string|null} body
 * @param {Date|null} signingTime
 * @returns {Record<string, string>}
 */
function signRequest(accessKey, accessSecret, appId, method, uriPath, rawQuery = '', body = null, signingTime = null) {
    method = method.toUpperCase();
    const now = signingTime || new Date();
    const tgDate = formatTgDate(now);
    const bodyStr = body || '';
    const payloadHash = sha256Hex(bodyStr);

    // Step 1: build header map
    const hv = {
        'X-Tg-Algorithm': ALG_TGV1,
        'X-Tg-Date': tgDate,
        'X-Tg-App-Id': appId.trim(),
        'X-Tg-Content-Sha256': payloadHash,
    };
    // Signed headers: only x-tg-app-id, x-tg-date (+ content-type, content-length for body methods)
    const signedNames = ['X-Tg-App-Id', 'X-Tg-Date'];
    if (['POST', 'PUT', 'PATCH'].includes(method)) {
        hv['Content-Type'] = 'application/json';
        hv['Content-Length'] = String(Buffer.byteLength(bodyStr, 'utf-8'));
        signedNames.unshift('Content-Length', 'Content-Type');
    }

    // Step 2: sorted signed header names
    const signedHdrs = signedNames.map(k => k.toLowerCase()).join(';');

    // Step 3: canonical headers string
    const lines = signedNames.map(k => `${k.toLowerCase()}:${hv[k]}`);
    const canonHdrs = lines.join('\n');

    // Step 4: canonical request
    const uri = canonicalURI(uriPath);
    const qCanon = canonicalQuery(method, rawQuery.replace(/^\?/, ''));
    const canonReq = [method, uri, qCanon, canonHdrs, signedHdrs, payloadHash].join('\n');

    // Step 5: string to sign
    const scopeDate = addDays(now, 7);
    const scope = `${scopeDate}/tgv1_request`;
    const strToSign = [ALG_TGV1, tgDate, scope, sha256Hex(canonReq)].join('\n');

    // Step 6: derive signing key
    let k = hmacSha256(formatDate(now), Buffer.from('TGV1' + accessSecret));
    k = hmacSha256(uri, k);
    k = hmacSha256('tgv1_request', k);

    // Step 7: signature
    const sig = hmacSha256Hex(strToSign, k);

    // Step 8: Authorization
    const auth = `${ALG_TGV1} Credential=${accessKey}/${scope}, SignedHeaders=${signedHdrs}, Signature=${sig}`;

    const out = {
        'X-Tg-Algorithm': hv['X-Tg-Algorithm'],
        'X-Tg-Date': hv['X-Tg-Date'],
        'X-Tg-App-Id': hv['X-Tg-App-Id'],
        'X-Tg-Content-Sha256': hv['X-Tg-Content-Sha256'],
        'X-Tg-Signed-Headers': signedHdrs,
        'Authorization': auth,
    };
    if (hv['Content-Type']) out['Content-Type'] = hv['Content-Type'];
    if (hv['Content-Length']) out['Content-Length'] = hv['Content-Length'];
    return out;
}

function sha256Hex(data) {
    return crypto.createHash('sha256').update(data, 'utf-8').digest('hex');
}

function hmacSha256(data, key) {
    return crypto.createHmac('sha256', key).update(data, 'utf-8').digest();
}

function hmacSha256Hex(data, key) {
    return crypto.createHmac('sha256', key).update(data, 'utf-8').digest('hex');
}

function canonicalURI(p) {
    p = p.trim();
    if (p.length > 1 && p.endsWith('/')) p = p.slice(0, -1);
    return p;
}

function canonicalQuery(method, rawQuery) {
    if (['POST', 'PUT', 'PATCH'].includes(method.toUpperCase())) return '';
    return rawQuery.replace(/\+/g, '%20');
}

function formatTgDate(d) {
    return d.toISOString().replace(/[-:]/g, '').replace(/\.\d{3}Z/, 'Z');
}

function formatDate(d) {
    return d.toISOString().slice(0, 10).replace(/-/g, '');
}

function addDays(d, n) {
    const r = new Date(d);
    r.setUTCDate(r.getUTCDate() + n);
    return formatDate(r);
}

module.exports = { signRequest };

File: tools/tirtc-api-client/python/tirtc_signing.py
"""
TGV1-HMAC-SHA256 request signing for tange.ai server APIs.

Usage:
    from tirtc_signing import sign_request

    headers = sign_request(
        access_key="...", access_secret="...", app_id="...",
        method="POST", uri_path="/v2/user/login/user-id",
        body='{"user_id":"test"}',
    )
"""

import hashlib
import hmac
from datetime import datetime, timedelta, timezone
from typing import Optional


ALG_TGV1 = "TGV1-HMAC-SHA256"


def sign_request(
    access_key: str,
    access_secret: str,
    app_id: str,
    method: str,
    uri_path: str,
    raw_query: str = "",
    body: Optional[str] = None,
    signing_time: Optional[datetime] = None,
) -> dict[str, str]:
    """Build signed HTTP headers for a tange.ai OpenAPI request."""
    method = method.upper()
    if signing_time is None:
        signing_time = datetime.now(timezone.utc)
    now = signing_time
    body_str = body or ""
    payload_hash = _sha256_hex(body_str)

    # Step 1: build header map
    hv = {
        "X-Tg-Algorithm": ALG_TGV1,
        "X-Tg-Date": now.strftime("%Y%m%dT%H%M%SZ"),
        "X-Tg-App-Id": app_id.strip(),
        "X-Tg-Content-Sha256": payload_hash,
    }
    # Signed headers: only x-tg-app-id, x-tg-date (+ content-type, content-length for body methods)
    signed_names = ["X-Tg-App-Id", "X-Tg-Date"]
    if method in ("POST", "PUT", "PATCH"):
        hv["Content-Type"] = "application/json"
        hv["Content-Length"] = str(len(body_str.encode("utf-8")))
        signed_names = ["Content-Length", "Content-Type"] + signed_names

    # Step 2: sorted signed header names
    signed_hdrs = ";".join(k.lower() for k in signed_names)
    lines = [f"{k.lower()}:{hv[k]}" for k in signed_names]
    canon_hdrs = "\n".join(lines)

    # Step 4: canonical request
    uri = _canonical_uri(uri_path)
    q_canon = _canonical_query(method, raw_query.lstrip("?"))
    canon_req = "\n".join([method, uri, q_canon, canon_hdrs, signed_hdrs, payload_hash])

    # Step 5: string to sign
    scope_date = (now + timedelta(days=7)).strftime("%Y%m%d")
    scope = f"{scope_date}/tgv1_request"
    str_to_sign = "\n".join([ALG_TGV1, hv["X-Tg-Date"], scope, _sha256_hex(canon_req)])

    # Step 6: derive signing key
    k = _hmac_sha256(now.strftime("%Y%m%d"), f"TGV1{access_secret}".encode())
    k = _hmac_sha256(uri, k)
    k = _hmac_sha256("tgv1_request", k)

    # Step 7: signature
    sig = _hmac_sha256_hex(str_to_sign, k)

    # Step 8: Authorization
    auth = f"{ALG_TGV1} Credential={access_key}/{scope}, SignedHeaders={signed_hdrs}, Signature={sig}"

    # Build output
    out = {
        "X-Tg-Algorithm": hv["X-Tg-Algorithm"],
        "X-Tg-Date": hv["X-Tg-Date"],
        "X-Tg-App-Id": hv["X-Tg-App-Id"],
        "X-Tg-Content-Sha256": hv["X-Tg-Content-Sha256"],
        "X-Tg-Signed-Headers": signed_hdrs,
        "Authorization": auth,
    }
    if "Content-Type" in hv:
        out["Content-Type"] = hv["Content-Type"]
    if "Content-Length" in hv:
        out["Content-Length"] = hv["Content-Length"]
    return out


def _sha256_hex(data: str) -> str:
    return hashlib.sha256(data.encode("utf-8")).hexdigest()


def _hmac_sha256(data: str, key: bytes) -> bytes:
    return hmac.new(key, data.encode("utf-8"), hashlib.sha256).digest()


def _hmac_sha256_hex(data: str, key: bytes) -> str:
    return hmac.new(key, data.encode("utf-8"), hashlib.sha256).hexdigest()


def _canonical_uri(p: str) -> str:
    p = p.strip()
    if len(p) > 1 and p.endswith("/"):
        p = p[:-1]
    return p


def _canonical_query(method: str, raw_query: str) -> str:
    if method in ("POST", "PUT", "PATCH"):
        return ""
    return raw_query.replace("+", "%20")
